compiler/goroutine-lowering.go

```diff
@@ -534,15 +532,12 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
 
 					// Allocate space for the return value.
 					var retvalAlloca llvm.Value
-					if inst.Type().TypeKind() != llvm.VoidTypeKind {
-						if retAlloc.IsNil() {
-							// insert at start of function
-							c.builder.SetInsertPointBefore(f.EntryBasicBlock().FirstInstruction())
+					if callee.Type().ElementType().ReturnType().TypeKind() != llvm.VoidTypeKind {
+						// insert at start of function
+						c.builder.SetInsertPointBefore(f.EntryBasicBlock().FirstInstruction())
 
-							// allocate return value buffer
-							retAlloc = c.builder.CreateAlloca(inst.Type(), "coro.retvalAlloca")
-						}
-						retvalAlloca = retAlloc
+						// allocate return value buffer
+						retvalAlloca = c.builder.CreateAlloca(callee.Type().ElementType().ReturnType(), "coro.retvalAlloca")
```

Comment:
Why not use [`createTemporaryAlloca`](https://github.com/tinygo-org/tinygo/blob/master/compiler/llvm.go#L49)? You would probably have to modify `createEntryBlockAlloca` to allow an insert point not at the end of a block, but that shouldn't be too difficult. By using these helpers, you get extra benefits like automatic lifetime calls (but don't forget to end the lifetime of the alloca). In theory that would allow the coro lowering passes to reuse stack slots in the coro frames, in practice that hasn't yet been implemented but does work around a LLVM bug in coroutine lowering.

Background: https://github.com/tinygo-org/tinygo/commit/7b6ef65fe7b927594e4f62cb7a7334137f4c2d9e